File: 2021_summer_algorithm_study_intermediate/using_python/2150_Tarjan.py
"""
input :
7 9
1 4
4 5
5 1
1 6
6 7
2 7
7 3
3 7
7 2

output :
3
1 4 5 -1
2 3 7 -1
6 -1
"""
import sys

sys.stdin = open("input.txt", "r")
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 6)


def dfs(s):
    global node_num
    global SCC_NUM
    node_num += 1
    dfn[s] = low[s] = node_num
    st.append(s)

    for nxt in graph[s]:
        if dfn[nxt] == 0:  # 처음 방문. 즉, tree edge
            dfs(nxt)
            low[s] = min(low[s], low[nxt])
        else:
            if finished[nxt]: continue

            low[s] = min(low[s], dfn[nxt])

    if dfn[s] == low[s]:
        SCC_NUM += 1
        while True:
            node = st.pop()
            finished[node] = True
            scc[node] = SCC_NUM

            if node == s:
                break



if __name__ == "__main__":
    V, E = map(int, input().rstrip().split())
    graph = [[] for _ in range(V + 1)]
    for _ in range(E):
        a, b = map(int, input().rstrip().split())
        graph[a].append(b)

    SCC_NUM = 0
    node_num = 0
    dfn = [0 for _ in range(V + 1)]
    low = [0 for _ in range(V + 1)]
    finished = [False for _ in range(V + 1)]
    scc = [0 for _ in range(V + 1)]
    st = []
    for i in range(1, V + 1):
        if finished[i]: continue
        dfs(i)

    res = [[] for _ in range(V + 1)]
    for i in range(1, V + 1):
        res[scc[i]].append(i)
    res.sort()

    print(SCC_NUM)
    res.sort()
    for group in res:
        if len(group) == 0: continue
        for node in group:
            print(node, end =' ')
        print(-1)

File: 2021_summer_algorithm_study_intermediate/using_cpp/2150_Tarjan.cpp
/*
input :
7 9
1 4
4 5
5 1
1 6
6 7
2 7
7 3
3 7
7 2

output :
3
1 4 5 -1
2 3 7 -1
6 -1
*/

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

// 전역 변수 선언
int V, E;
int SCC_NUM;
int node_num;
int a, b;

vector<int> graph[10005];
int dfn[10005];
int low[10005];
bool finished[10005];
int scc[10005];
vector<int> st;
vector<vector<int>> ans;


bool comp(vector<int> a, vector<int> b){
    return a[0] < b[0];
}


void dfs(int s){
    dfn[s] = low[s] = ++node_num;
    st.push_back(s);

    for (int nxt: graph[s]){
        if (dfn[nxt] == 0){
            dfs(nxt);
            low[s] = min(low[s], low[nxt]);
        }
        else{
            if (finished[nxt]) continue;
            low[s] = min(low[s], dfn[nxt]);
        }
    }


    if (dfn[s] == low[s]){
        SCC_NUM++;

        vector<int> tmp;
        while (true){
            int nxt = st.back(); st.pop_back();
            tmp.push_back(nxt);
            scc[nxt] = SCC_NUM;
            finished[nxt] = true;

            if (nxt == s) break;
        }
        sort(tmp.begin(), tmp.end());
        ans.push_back(tmp);
    }
}


int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);

    cin >> V >> E;
    while(E--){
        cin >> a >> b;
        graph[a].push_back(b);
    }

    for (int i = 1; i <= V; i++){
        if (finished[i]) continue;

        dfs(i);
    }
    sort(ans.begin(), ans.end(), comp);

    cout << SCC_NUM << '\n';
    for (vector<int> nodes: ans){
        for (int i = 0; i < nodes.size(); i++){
            cout << nodes[i] << ' ';
        }
        cout << "-1\n";
    }
    return 0;
}